ntbs-service/DataAccess/UserRepository.cs

```diff
@@ -0,0 +1,85 @@
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.EntityFrameworkCore;
+using ntbs_service.Models.Entities;
+using ntbs_service.Models.ReferenceEntities;
+
+namespace ntbs_service.DataAccess
+{
+    public interface IUserRepository
+    {
+        Task AddOrUpdateUser(User user, IEnumerable<TBService> tbServices);
+    }
+
+    public class UserRepository : IUserRepository
+    {
+        private readonly NtbsContext _context;
+
+        public UserRepository(NtbsContext context)
+        {
+            _context = context;
+        }
+
+        public async Task AddOrUpdateUser(User user, IEnumerable<TBService> tbServices)
+        {
+            var existingUser = await _context.User.Include(u => u.CaseManagerTbServices)
+                .SingleOrDefaultAsync(u => u.Username == user.Username);
+
+            if (existingUser != null)
+            {
+                await UpdateUser(existingUser, user, tbServices);
+            }
+            else
+            {
+                await AddUser(user, tbServices);
+            }
+        }
+
+        private async Task AddUser(User user, IEnumerable<TBService> tbServices)
+        {
+            AddCaseManagerTbServices(user, tbServices);
+            _context.User.Add(user);
+            await _context.SaveChangesAsync();
+        }
+
+        private async Task UpdateUser(User existingUser, User newUser, IEnumerable<TBService> tbServices)
+        {
+            DetachLocalTrackedObject(existingUser);
+
+            existingUser = newUser;
+            AddCaseManagerTbServices(existingUser, tbServices);
+            _context.Entry(existingUser).State = EntityState.Modified;
+
+            await _context.SaveChangesAsync();
+        }
+
+        private void DetachLocalTrackedObject(User existingUser)
+        {
+            // Not sure why this is necessary/what is tracking the same entity twice, but without kept getting error:
+            // "The instance of entity type cannot be tracked because another instance of this type with the same key [Username] is already being tracked"
+            // Solution from https://stackoverflow.com/questions/36856073/the-instance-of-entity-type-cannot-be-tracked-because-another-instance-of-this-t
+            // TODO: Work out if there is a nicer way.

```

Comment:
really weird... not sure why but the context seems to add the same username multiple times for tracking!

Reply:
I think the way it works is:
- we are in the case where w're updating the user, rather adding a new one
- `existingUser` is bound to the context
- `newUser` is another instance of a `User` object that is not bound to the context
- they share their primary key
- running `existingUser = newUser` only does variable reassignment, so `existingUser` now actually points to an object containing the new values but no bound to the context.
- running `_context(existingUser)` attempts to bind it to the context, but there's a clash of PKs!

I think we can avoid it by copying over the values from `newUser` to `existingUser`, much like we do in the case of notifications, so something along the lines of:
```
private async Task UpdateUser(User existingUser, User newUser, IEnumerable<TBService> tbServices)
        {
            _context.Entry(existingUser).CurrentValues.SetValues(newUser);
            AddCaseManagerTbServices(existingUser, tbServices);
            await _context.SaveChangesAsync();
        }
```